keps/sig-windows/windows-runtimeclass-support.md

```diff
@@ -0,0 +1,555 @@
+    scheduling:
+      nodeSelector:
+        kubernetes.io/os: 'windows'
+        kubernetes.io/arch: 'amd64'
+        node.kubernetes.io/osversion: '10.0.18362'
+      tolerations:
+      - effect: NoSchedule
+        key: windows
+        operator: Equal
+        value: "true"
+    ```
+
+Another RuntimeClass could still be run on the same hosts to use updated containers without Hyper-V isolation.
+
+    ```yaml
+    apiVersion: node.k8s.io/v1beta1
+    kind: RuntimeClass
+    metadata:
+      name: windows-1903
+    handler: 'default'
+    scheduling:
+      nodeSelector:
+        kubernetes.io/os: 'windows'
+        kubernetes.io/arch: 'amd64'
+        node.kubernetes.io/osversion: '10.0.18362'
+      tolerations:
+      - effect: NoSchedule
+        key: windows
+        operator: Equal
+        value: "true"
+    ```
+
+#### Story 3 - Choosing a specific multi-arch image
+
+Starting from story 2, the RuntimeClass `handler` field could also be used as a means os/version/architecture/variant used by the container runtime. The `handler` is matched up with a corresponding section in the ContainerD configuration file.
+
+Here's an example of what corresponding ContainerD configurations could look like for the runtimes above. The first one would force Hyper-V for compatibility with containers needing 10.0.17763. The second one would be the default for the current version where no Hyper-V isolation is required for compatibility.
+
+```toml
+        [plugins.cri.containerd.runtimes.containerd-hyperv-17763]
+          runtime_type = "io.containerd.runhcs.v1"
+          [plugins.cri.containerd.runtimes.containerd-hyperv-17763.options]
+            SandboxImage = "{{WINDOWSSANDBOXIMAGE}}"
+            SandboxPlatform = "windows/amd64"
+            SandboxOsVersion = "10.0.17763"
+            SandboxIsolation = 1
+
+        [plugins.cri.containerd.runtimes.default]
+          runtime_type = "io.containerd.runhcs.v1"
+          # No version is specified for process isolation, the node OS version is used
+```
+
+
+### Implementation Details/Notes/Constraints
+
+There were multiple options discussed with SIG-Node & SIG-Windows on October 8 2019 prior to filing this KEP. That discussion & feedback were captured in [Difficulties in mixed OS and arch clusters]. If you're looking for more details on other approaches excluded, please review that document.
+
+
+#### Adding new label node.kubernetes.io/osversion
+
+In [Bounding Self-Labeling Kubelets], a specific range of prefixes were reserved for node self-labeling - `[*.]node.kubernetes.io/*`. Adding a new label within that namespace won't require any changes to NodeRestriction admission. As a new field it also won't require changes to any existing workloads.
+
+Windows product names such as "Windows Server version 1903" don't align with the actual build numbers, so build numbers will be used instead. Users can choose to reflect the product name instead as shown in the user stories.
```

Comment:
This is a bummer. Build number versioning has always felt obscure to me; as a user I know definitively that I'm running "Windows Server version 1903" for example, and if I need to translate that to a build number that's a hassle.

I'll also point out that Microsoft [tags its base containers](https://hub.docker.com/_/microsoft-windows-servercore) with e.g. `ltsc2019` and `1903`; there would be value in making the `osversion` match the tags users will use for their containers. Is there no way to make this happen?

Reply:
The difficulty we have here is that the marketing names (Windows Server version 1903) change in the weeks leading up to a final release and are not clearly returned in APIs. There's also the chance that multiple OS builds - for example Windows Server 2019 & Windows Server version 1809 will be the same code with different marketing names confusing matters further. 

Testing insider builds also presents a challenge with marketing names because they have a placeholder name "Windows Server vNext". There's no compatibility between prerelease builds, so if you're testing [18975](https://blogs.windows.com/windowsexperience/2019/09/10/announcing-windows-server-vnext-insider-preview-build-18975/) and [18995](https://techcommunity.microsoft.com/t5/Windows-Server-Insiders/Announcing-Windows-Server-vNext-Preview-Build-18995/m-p/900354) then you need to be more specific.

`os.version` is part of the container manifest today and is set on all Windows images already based on a version number. My proposal is to align the OCI manifest usage with the kubelet label so that if/when marketing names are changed - Kubernetes doesn't need a code change to update a map of build numbers -> marketing names.

Reply:
I added some more details on this in the latest update. There's a chart of product names -> build numbers